File: inflearn/inflearn-spring-security-junit-bank-class/src/test/java/shop/mtcoding/bank/service/user/UserServiceTest.java
package shop.mtcoding.bank.service.user;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import shop.mtcoding.bank.config.dummy.DummyObject;
import shop.mtcoding.bank.domain.user.User;
import shop.mtcoding.bank.domain.user.UserRepository;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static shop.mtcoding.bank.dto.user.UserRequest.JoinRequest;
import static shop.mtcoding.bank.dto.user.UserResponse.JoinResponse;

@ExtendWith(MockitoExtension.class)
class UserServiceTest extends DummyObject {

    @InjectMocks
    private UserService userService;

    @Mock
    private UserRepository userRepository;

    @Spy
    private BCryptPasswordEncoder passwordEncoder;

    @Test
    void signUpTest() {
        // given
        JoinRequest joinRequest = new JoinRequest();
        joinRequest.setUsername("soono");
        joinRequest.setPassword("1234");
        joinRequest.setEmail("shyoon991@gmail.com");
        joinRequest.setFullname("soono");

        // stub - 1
        when(userRepository.findByUsername(any())).thenReturn(Optional.empty());

        // stub - 2
        User user = newMockUser(1L, "soono", "soono");
        when(userRepository.save(any())).thenReturn(user);

        // when
        JoinResponse joinResponse = userService.signUp(joinRequest);

        // then
        assertThat(joinResponse.getId()).isEqualTo(1L);
        assertThat(joinResponse.getUsername()).isEqualTo("soono");
    }

}

File: src/test/kotlin/com/group/libraryapp/service/user/UserServiceTest.kt
package com.group.libraryapp.service.user

import com.group.libraryapp.domain.user.User
import com.group.libraryapp.domain.user.UserRepository
import com.group.libraryapp.dto.user.request.UserCreateRequest
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class UserServiceTest @Autowired constructor(
    private val userRepository: UserRepository,
    private val userService: UserService
) {

    // 각 테스트가 종료할 때마다 데이터를 초기화 해준다.
    @AfterEach
    fun clean() {
        userRepository.deleteAll()
    }

    @Test
    fun saveUserTest() {
        // given
        val request = UserCreateRequest("soono", null)

        // when
        userService.saveUser(request)

        // then
        val results = userRepository.findAll()
        assertThat(results).hasSize(1)
        assertThat(results[0].name).isEqualTo("soono")
        assertThat(results[0].age).isNull()
    }

    @Test
    fun getUsersTest() {
        // given
        userRepository.saveAll(
            listOf(
                User("A", 20),
                User("B", null)
            )
        )

        // when
        val results = userService.getUsers()

        // then
        assertThat(results).hasSize(2) // [UserResponse(), UserResponse()
        assertThat(results)
            .extracting("name") // ["A", "B"]
            .containsExactlyInAnyOrder("A", "B")

        assertThat(results)
            .extracting("age") // [20, null]
            .containsExactlyInAnyOrder(20, null)
    }

}